File: LeetCode/3/3.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        maxLen=0
        start=0
        charSet=set()

        for end in range(len(s)):
            if s[end] not in charSet:
                charSet.add(s[end])
                maxLen = max(maxLen, end-start+1)
            else:
                while s[end] in charSet:
                    charSet.remove(s[start])
                    start=start+1
                charSet.add(s[end])
        
        return maxLen

        

File: LeetCode/3/3.cpp
class Solution {
public:
    int lengthOfLongestSubstring(std::string s) {
        std::unordered_map<char, int> lastSeen;
        int maxLen = 0;
        int start = 0;

        for (int end = 0; end < s.length(); ++end) {
            char c = s[end];
            if (lastSeen.count(c) && lastSeen[c] >= start) {
                start = lastSeen[c] + 1;
            }
            lastSeen[c] = end;
            maxLen = std::max(maxLen, end - start + 1);
        }

        return maxLen;
    }
};